File: packages/ios/Sources/Components/Rating/Rating.swift
import SwiftUI

/// What a rating is measured on.
///
/// `stars` is an ordinal scale: choosing four means "at least four", so four
/// fill. `thumbs` is a choice between two, and exactly the one chosen fills —
/// a thumbs-up is not "two thumbs". The value is a number either way:
/// **0 is unanswered, 1 is down, 2 is up**, the same "1 is the lowest" rule
/// the stars follow.
public enum KozmosRatingVariant: Sendable, Hashable, CaseIterable {
    case stars
    case thumbs
}

/// Mirrors the React `Rating` and Compose's `KozmosRating`.
///
/// Every option is a Button. It was an `Image` with `.onTapGesture`, which
/// VoiceOver cannot activate and does not announce: the rating was invisible
/// to it, and unreachable by Switch Control and a keyboard.
public struct KozmosRating: View {
    @Binding var value: Int
    let variant: KozmosRatingVariant
    let max: Int
    let readOnly: Bool
    let label: String
    let itemLabel: (Int, Int, KozmosRatingVariant) -> String
    let valueLabel: (Int, Int, KozmosRatingVariant) -> String

    public init(
        value: Binding<Int>,
        variant: KozmosRatingVariant = .stars,
        max: Int = 5,
        readOnly: Bool = false,
        label: String = "Rating",
        itemLabel: @escaping (Int, Int, KozmosRatingVariant) -> String = KozmosRating.defaultItemLabel,
        valueLabel: @escaping (Int, Int, KozmosRatingVariant) -> String = KozmosRating.defaultValueLabel
    ) {
        self._value = value
        self.variant = variant
        self.max = variant == .thumbs ? 2 : max
        self.readOnly = readOnly
        self.label = label
        self.itemLabel = itemLabel
        self.valueLabel = valueLabel
    }

    public static func defaultItemLabel(_ item: Int, _ max: Int, _ variant: KozmosRatingVariant) -> String {
        variant == .thumbs ? (item == 1 ? "Poor" : "Good") : "Rate \(item) out of \(max) stars"
    }

    public static func defaultValueLabel(_ value: Int, _ max: Int, _ variant: KozmosRatingVariant) -> String {
        if value == 0 { return "Not rated" }
        return variant == .thumbs
            ? (value == 1 ? "Rated poor" : "Rated good")
            : "Rated \(value) out of \(max) stars"
    }

    private func filled(_ item: Int) -> Bool {
        variant == .thumbs ? item == value : item <= value
    }

    public var body: some View {
        HStack(spacing: KozmosDimensions.primitivesLayoutSpacing50) {
            ForEach(1...max, id: \.self) { item in
                if readOnly {
                    face(item)
                } else {
                    Button {
                        // Choosing what is already chosen clears it, so a
                        // visitor who taps the wrong thumb can undo it without
                        // dismissing the whole dialog.
                        withAnimation { value = (item == value) ? 0 : item }
                    } label: {
                        face(item)
                    }
                    .buttonStyle(.plain)
                    .accessibilityLabel(itemLabel(item, max, variant))
                    .accessibilityAddTraits(item == value ? [.isButton, .isSelected] : .isButton)
                }
            }
        }
        // Read-only is not a control: it reads as one image with the rating as
        // its label, rather than as a row of things that cannot be pressed.
        .accessibilityElement(children: readOnly ? .ignore : .contain)
        .accessibilityLabel(readOnly ? valueLabel(value, max, variant) : label)
    }

    @ViewBuilder private func face(_ item: Int) -> some View {
        switch variant {
        case .stars:
            Image(systemName: filled(item) ? "star.fill" : "star")
                .foregroundColor(
                    filled(item)
                        ? KozmosColors.semanticsDataYellow
                        : KozmosColors.primitivesColorsForeground400
                )
                .font(KozmosTypography.title2)
        case .thumbs:
            Image(systemName: item == 1 ? "hand.thumbsdown" : "hand.thumbsup")
                .font(KozmosTypography.title3)
                .foregroundColor(
                    filled(item)
                        ? KozmosColors.primitivesColorsTheme600
                        : KozmosColors.primitivesColorsForeground400
                )
                .frame(width: 40, height: 40)
                .background(
                    Circle().fill(
                        filled(item)
                            ? KozmosColors.primitivesColorsTheme0
                            : KozmosColors.primitivesColorsBackground100
                    )
                )
                .overlay(
                    Circle().stroke(
                        filled(item) ? KozmosColors.primitivesColorsTheme600 : .clear,
                        lineWidth: 2
                    )
                )
                // 44 is the target even though the circle is 40.
                .frame(width: 44, height: 44)
        }
    }
}

File: packages/android/src/main/java/com/kozmos/components/Rating/Rating.kt
package com.kozmos.components.rating

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Star
import androidx.compose.material.icons.filled.ThumbDown
import androidx.compose.material.icons.filled.ThumbUp
import androidx.compose.material.icons.outlined.Star
import androidx.compose.material3.Icon
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp
import com.kozmos.tokens.KozmosDimensions
import com.kozmos.tokens.KozmosThemeTokens

/**
 * What a rating is measured on.
 *
 * [Stars] is an ordinal scale: choosing four means "at least four", so four
 * fill. [Thumbs] is a choice between two, and exactly the one chosen fills - a
 * thumbs-up is not "two thumbs". The value is a number either way: **0 is
 * unanswered, 1 is down, 2 is up**, the same "1 is the lowest" rule the stars
 * follow.
 */
enum class KozmosRatingVariant {
    Stars,
    Thumbs
}

/**
 * Mirrors the React `Rating` and SwiftUI's `KozmosRating`.
 *
 * Every option is `selectable` with [Role.RadioButton] and a description.
 * Before, each was an `Icon` with `contentDescription = null` and a bare
 * `clickable`: TalkBack announced nothing at all, so the rating existed only
 * for people using their eyes and a finger.
 */
@Composable
fun KozmosRating(
    value: Int,
    onValueChange: (Int) -> Unit,
    modifier: Modifier = Modifier,
    variant: KozmosRatingVariant = KozmosRatingVariant.Stars,
    max: Int = 5,
    readOnly: Boolean = false,
    label: String = "Rating",
    itemLabel: (Int, Int, KozmosRatingVariant) -> String = ::kozmosRatingItemLabel,
    valueLabel: (Int, Int, KozmosRatingVariant) -> String = ::kozmosRatingValueLabel
) {
    val thumbs = variant == KozmosRatingVariant.Thumbs
    val steps = if (thumbs) 2 else max
    val filled: (Int) -> Boolean = { item -> if (thumbs) item == value else item <= value }

    // Read-only is not a control: it reads as one thing whose description is
    // the rating, rather than as a row of options nobody can take.
    val rowModifier = if (readOnly) {
        modifier.clearAndSetSemantics { contentDescription = valueLabel(value, steps, variant) }
    } else {
        modifier.semantics { contentDescription = label }.selectableGroup()
    }

    Row(
        modifier = rowModifier,
        // The stars space themselves: each glyph sits in a 32 box with 2 of
        // padding, so there are already 4 between them — the web's gap-1 and
        // SwiftUI's spacing50. Adding a row gap on top made it 8, which is a
        // silent redesign smuggled in behind an accessibility fix. The thumbs
        // are two circles and need the gap.
        horizontalArrangement =
            if (thumbs) Arrangement.spacedBy(KozmosDimensions.primitivesLayoutSpacing100)
            else Arrangement.Start,
        verticalAlignment = Alignment.CenterVertically
    ) {
        repeat(steps) { index ->
            val item = index + 1
            val itemModifier = if (readOnly) {
                Modifier
            } else {
                Modifier.selectable(
                    selected = item == value,
                    role = Role.RadioButton,
                    onClick = {
                        // Choosing what is already chosen clears it.
                        onValueChange(if (item == value) 0 else item)
                    }
                ).semantics { contentDescription = itemLabel(item, steps, variant) }
            }
            Box(modifier = itemModifier, contentAlignment = Alignment.Center) {
                if (thumbs) {
                    Box(
                        modifier = Modifier
                            .size(40.dp)
                            .background(
                                if (filled(item)) KozmosThemeTokens.primitivesColorsTheme0
                                else KozmosThemeTokens.primitivesColorsBackground100,
                                CircleShape
                            )
                            .border(
                                2.dp,
                                if (filled(item)) KozmosThemeTokens.primitivesColorsTheme600
                                else Color.Transparent,
                                CircleShape
                            ),
                        contentAlignment = Alignment.Center
                    ) {
                        Icon(
                            imageVector = if (item == 1) Icons.Filled.ThumbDown else Icons.Filled.ThumbUp,
                            contentDescription = null,
                            tint = if (filled(item)) KozmosThemeTokens.primitivesColorsTheme600
                            else KozmosThemeTokens.primitivesColorsForeground400,
                            modifier = Modifier
                            .size(KozmosDimensions.primitivesLayoutSizing300)
                            .padding(KozmosDimensions.primitivesLayoutSpacing25)
                        )
                    }
                } else {
                    Icon(
                        imageVector = if (filled(item)) Icons.Filled.Star else Icons.Outlined.Star,
                        contentDescription = null,
                        // semanticsDataYellow, as React and SwiftUI draw it.
                        // This was primitivesColorsEmotionalAlert600 - #f9a707
                        // against the other two platforms' #d97706, so the same
                        // star was a different colour on Android, in both themes.
                        tint = if (filled(item)) KozmosThemeTokens.semanticsDataYellow
                        else KozmosThemeTokens.primitivesColorsForeground400,
                        // 32 box, 2 of padding: a 28 glyph, exactly the size
                        // it was. Only the colour and the semantics change
                        // here — dropping the padding would have grown every
                        // star by four and called it an accessibility fix.
                        modifier = Modifier
                            .size(KozmosDimensions.primitivesLayoutSizing400)
                            .padding(KozmosDimensions.primitivesLayoutSpacing25)
                    )
                }
            }
        }
    }
}

fun kozmosRatingItemLabel(item: Int, max: Int, variant: KozmosRatingVariant): String =
    if (variant == KozmosRatingVariant.Thumbs) {
        if (item == 1) "Poor" else "Good"
    } else {
        "Rate $item out of $max stars"
    }

fun kozmosRatingValueLabel(value: Int, max: Int, variant: KozmosRatingVariant): String =
    when {
        value == 0 -> "Not rated"
        variant == KozmosRatingVariant.Thumbs -> if (value == 1) "Rated poor" else "Rated good"
        else -> "Rated $value out of $max stars"
    }
